src/ripple/app/main/Application.cpp

```diff
@@ -352,6 +352,8 @@ class ApplicationImp : public Application, public RootStoppable, public BasicApp
     detail::AppFamily family_;
     std::unique_ptr<NodeStore::DatabaseShard> shardStore_;
     std::unique_ptr<detail::AppFamily> shardFamily_;
+    std::unique_ptr<RPC::ShardArchiveHandler> shardArchiveHandler_;
+    std::mutex shardArchiveHandlerMutex_;
```

Comment:
I'm not sure I see the race condition you mentioned.

Reply:
If we make `shardArchiveHandler_` a static member, this makes it difficult to ensure that the `Application` is constructing a fresh instance when creating the handler at program startup (due to preexisting state).

Reply:
Sorry, I didn't make myself clear.  I'm not suggesting that `ShardArchiveHandler_` be a static.

I was suggesting that the `std::mutex`, `shardArchiveHandlerMutex_` could be made a function-local `static` inside of `getShardArchiveHandler()`.  That way the scope is limited to just that function.  It's the easiest possible to think about `mutex`; by implementation, only usable inside that one function.

But that change can only work if we agree that member function `hasShardArchiveHandler()` is not useful, since that's the only place outside of `getShardArchiveHandler()` that uses `shardArchiveHandlerMutex_`.

Reply:
I'd rather keep `hasShardArchiveHandler` so that unit tests can verify that interrupted downloads are resumed. This enables us to use unit tests (rather than verification testing) to confirm a major feature that was added in the previous PR.

Reply:
Um, okay.  Be aware that you're adding an entry point to the `Application` (which has very wide visibility) that is only being used for one unit test of what most folks would consider to be a minor feature.  You gotta do what feels right for you.  But, personally, I don't think adding an entry point in the `Application` is worth it.

It sounds to me like what you feel is really important to know is which call to `getShardArchiveHandler` actually constructed the `ShardArchiveHandler`.  Maybe there's a way to achieve that verification without adding a test stub to the `Application`.  Here's a thought, it's not necessarily great, but maybe you'll come up with a better idea.

Suppose we were to add a defaulted integer argument to
```
getShardArchiveHandler(int sig = 0)
```
That integer is in turn passed to the `ShardArchiveHandler` constructor as an argument.  The `ShardArchiveHandler` tucks the value of the argument away as a `const` data member.

The only thing that will look at the tucked away integer is your tests.  But since your tests are `friend`s, they can look directly at the value.  The `ShardArchiveHandler` doesn't even need to provide an accessor.

Any non-test callers of `getShardArchiveHandler()` ignore the argument, so for them the value is always defaulted to zero.  However in your tests you can pass non-zero arguments.  This will allow you to determine whether the `ShardArchiveHandler` was constructed by your test or by something else in the application.

Taking this approach your `TestStateDatabase4` test can call `getShardArchiveHandler(42)`.  If the returned `ShardArchiveHandler` contains `0`, then you know that you didn't construct it.  It must have been constructed before your call to `getShardArchiveHandler()`, so your feature is working.

If the returned `ShardArchiveHandler` contains `42` then your feature is not working.

I think this approach is better because the testability features are (mostly) just being added to the `ShardArchiveHandler`, not to the `Application`.

Not pretty.  Perhaps you have a better idea?

Reply:
What about an empty subclass called `RecoveryHandler` (tentative name) that inherits from `ShardArchiveHandler`. The application constructs a `RecoveryHandler` if it detects preexisting state. Then the unit test can use dynamic pointer casts to confirm the type.

Reply:
Sounds like a good idea.  Thanks for your efforts continuing to push on this.